### PR TITLE
Refactor how optimized methods for Gaudi replace their counterparts in Transformers

### PR DESCRIPTION
**Pull request by regisss**
# What does this PR do?

Some forward methods (e.g. for ViT and ALBERT) and generation methods are tweaked so that they are optimized for running on Gaudi. The replacement logic with their Transformers' counterparts is currently done in the Trainer. This PR moves this logic in `modeling_utils.py` so that only one call is needed in the Trainer.

The way these methods are replaced is not the most elegant one, but it will be much easier to maintain and will profit all models from Transformers (not only the one we validated). For instance, before this PR, only GPT2 and T5 benefit from the optimized generation methods. With this PR, all the models that have a `generate` method will benefit from them.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/habana/models/modeling_all_models.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    if encoder_attention_mask.dim() == 3:
+        encoder_extended_attention_mask = encoder_attention_mask[:, None, :, :]
+    if encoder_attention_mask.dim() == 2:
+        encoder_extended_attention_mask = encoder_attention_mask[:, None, None, :]
+    # T5 has a mask that can compare sequence ids, we can simulate this here with this transposition
+    # Cf. https://github.com/tensorflow/mesh/blob/8d2465e9bc93129b913b5ccc6a59aa97abd96ec6/mesh_tensorflow
+    # /transformer/transformer_layers.py#L270
+    # encoder_extended_attention_mask = (encoder_extended_attention_mask ==
+    # encoder_extended_attention_mask.transpose(-1, -2))
+    # HMP is disabled because (1.0 - encoder_extended_attention_mask) may be converted into bf16
```
[the remark this answers is not in the document]
Comment (regisss) [reply optimum/habana/models/modeling_all_models.py:24]: Yes, which makes sense since `rsub` is in the list of bf16 ops.

**Review by libinta — APPROVED** [review]
@regisss normally we prefer model file not change but changing the config file. But case by case.

File: optimum/habana/models/modeling_all_models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (regisss) [reply optimum/habana/models/modeling_all_models.py:24]: Well, we could remove `rsub` from the bf16 list, but then we may get lower throughput and higher TTT.

**Comment by regisss** [comment]
> @regisss normally we prefer model file not change but changing the config file. But case by case.

I understand, but since this is used by all NLP models in Transformers, that would basically mean that `rsub` should never be in the list of bf16 ops. IMO the current solution enables to keep the best of Transformers and HMP.

**Pull request merged**